graphium/graphium_cpp/features.cpp

```diff
@@ -0,0 +1,1395 @@
+        uint32_t a = bonds[i].beginAtomIdx;
+        uint32_t b = bonds[i].endAtomIdx;
+        // NOTE: +1 is because first entry will stay zero.
+        ++neighbour_starts[a + 1];
+        ++neighbour_starts[b + 1];
+    }
+
+    // Find the starts by partial-summing the neighbour counts.
+    // NOTE: +1 is because first entry will stay zero.
+    std::partial_sum(neighbour_starts + 1, neighbour_starts + 1 + num_atoms, neighbour_starts + 1);
+
+    // Fill in the neighbours and bond_indices arrays.
+    uint32_t* neighbours = neighbour_starts + num_atoms + 1;
+    uint32_t* bond_indices = neighbours + 2 * num_bonds;
+    for (uint32_t i = 0; i < num_bonds; ++i) {
+        uint32_t a = bonds[i].beginAtomIdx;
+        uint32_t b = bonds[i].endAtomIdx;
+
+        uint32_t ai = neighbour_starts[a];
+        neighbours[ai] = b;
+        bond_indices[ai] = i;
+        ++neighbour_starts[a];
+
+        uint32_t bi = neighbour_starts[b];
+        neighbours[bi] = a;
+        bond_indices[bi] = i;
+        ++neighbour_starts[b];
+    }
+
+    // Shift neighbour_starts forward one after incrementing it.
+    uint32_t previous = 0;
+    for (uint32_t i = 0; i < num_atoms; ++i) {
+        uint32_t next = neighbour_starts[i];
+        neighbour_starts[i] = previous;
+        previous = next;
+    }
+
+    // NeighbourData takes ownership of the memory.
+    return NeighbourData{ std::move(deleter), neighbour_starts, neighbours, bond_indices };
+}
+
+// This fills in 3 values for each atom
+template<typename T>
+at::Tensor get_conformer_features(
+    const RDKit::ROMol &mol,
+    c10::ScalarType dtype,
+    MaskNaNStyle mask_nan_style,
+    T mask_nan_value,
+    int64_t &num_nans) {
+
+    const size_t n = mol.getNumAtoms();
+    std::unique_ptr<T[]> conformer_data(new T[3 * n]);
+    T* data = conformer_data.get();
+
+    if (mol.beginConformers() == mol.endConformers()) {
+        // No conformers: treat as NaN
+        if (mask_nan_style == MaskNaNStyle::REPORT) {
+            num_nans += 3*n;
+        }
+        else {
+            for (size_t i = 0; i < 3 * n; ++i) {
+                data[i] = mask_nan_value;
+            }
+        }
+    }
+    else {
+        const RDKit::Conformer& conformer = mol.getConformer();
+        const auto& positions = conformer.getPositions();
+        assert(positions.size() == n);
+        for (size_t i = 0; i < n; ++i, data += 3) {
+            const auto& position = positions[i];
+            data[0] = FeatureValues<T>::convertToFeatureType(position.x);
+            data[1] = FeatureValues<T>::convertToFeatureType(position.y);
+            data[2] = FeatureValues<T>::convertToFeatureType(position.z);
+        }
+
+        num_nans += mask_nans(data, 3 * n, mask_nan_style, mask_nan_value);
+    }
+
+    const int64_t dims[1] = { int64_t(3 * n) };
+    return torch_tensor_from_array<T>(std::move(conformer_data), dims, 1, dtype);
+}
+
+static const std::unordered_map<std::string, int64_t> atom_float_name_to_enum{
```

Comment:
This mapping is for converting strings to enum values, not floats, so for example, if `atom_float_feature_names_to_tensor` receives `["atomic-number", "mass", "chirality"]`, it will return a Tensor containing the integers `[0, 1, 5]`.  Then, when featurizing, `featurize_smiles` receives the integer Tensor, instead of the strings, and uses the enum values to determine what features are needed.  This is done when preparing the data module, to reduce the amount of string processing that needs to be done at featurization time.

Reply:
Ahh yeah, I'm a bit rusty in C++.